scripts/vim-patch.sh

```diff
@@ -26,7 +26,28 @@ usage() {
   echo "The default is '${VIM_SOURCE_DIR_DEFAULT}'."
 }
 
+# Checks if a program is in the user's PATH, and is executable.
+check_executable() {
+  if [[ ! -x $(command -v "${1}") ]]; then
+    >&2 echo "${BASENAME}: '${1}' not found in PATH or not executable."
+    >&2 echo "Aborting execution."
+    exit 1
+  fi
+}
+
+# Checks for the right number of parameters.
+check_params() {
```

Comment:
I converted it to getopts, and I think it turned out fine. We lose the long options, but we don't have many options so I don't think it's a big deal, but we do get explicit errors if the user a) doesn't provide an argument to an option b) provides an unknown option argument.

Also, now that we have two different modes (review mode, and patch mode), I thought it best to just remove the positional arguments in favor of explicit option arguments.

```diff
diff --git a/scripts/vim-patch.sh b/scripts/vim-patch.sh
index dd28b2b..10c9d42 100755
--- a/scripts/vim-patch.sh
+++ b/scripts/vim-patch.sh
@@ -11,16 +11,15 @@ usage() {
   echo "Helper script for porting Vim patches. For more information, see"
   echo "https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-vim"
   echo
-  echo "Usage:  ${BASENAME} [option]"
-  echo "        ${BASENAME} vim-revision"
+  echo "Usage:  ${BASENAME} [-h | -l | -p vim-revision | -r pr-number]"
   echo
   echo "Options:"
-  echo "    -h, --help                Show this message."
-  echo "    -l, --list                Show list of Vim patches missing from Neovim."
-  echo "    -r, --review {pr-number}  Review a vim-patch pull request to Neovim."
-  echo
-  echo "vim-revision can be a version number of the format '7.4.xxx'"
-  echo "or a Mercurial commit hash."
+  echo "    -h                 Show this message and exit."
+  echo "    -l                 Show list of Vim patches missing from Neovim."
+  echo "    -p {vim-revision}  Download and apply the Vim patch vim-revision."
+  echo "                       vim-revision can be a version number of the "
+  echo "                       format '7.4.xxx' or a Mercurial commit hash."
+  echo "    -r {pr-number}     Review a vim-patch pull request to Neovim."
   echo
   echo "Set VIM_SOURCE_DIR to change where Vim's sources are stored."
   echo "The default is '${VIM_SOURCE_DIR_DEFAULT}'."
@@ -35,16 +34,6 @@ check_executable() {
   fi
 }
 
-# Checks for the right number of parameters.
-check_params() {
-  # Add one because the expected number of parameters
-  # is also passed to this function.
-  if [[ ${#} != $((${1}+1)) ]]; then
-    usage
-    exit 1
-  fi
-}
-
 get_vim_sources() {
   check_executable hg
 
@@ -274,18 +263,30 @@ review_pr() {
   exec nvim -O "${NEOVIM_SOURCE_DIR}/${base_name}.diff" "${NEOVIM_SOURCE_DIR}/n${base_name}.diff"
 }
 
-if [[ ${#} == 0 || ${1} == "--help" || ${1} == "-h" ]]; then
-  usage
-  exit 0
-fi
+while getopts "hlp:r:" opt; do
+  case $opt in
+    h)
+      usage
+      exit 0
+      ;;
+    l)
+      list_vim_patches
+      exit 0
+      ;;
+    p)
+      get_vim_patch "$OPTARG"
+      exit 0
+      ;;
+    r)
+      review_pr "$OPTARG"
+      exit 0
+      ;;
+    *)
+      exit 1
+      ;;
+  esac
+done
 
-if [[ ${1} == "--list" || ${1} == "-l" ]]; then
-  check_params 1 "${@}"
-  list_vim_patches
-elif [[ ${1} == "--review" || ${1} == "-r" ]]; then
-  check_params 2 "${@}"
-  review_pr "${2}"
-else
-  check_params 1 "${@}"
-  get_vim_patch "${1}"
-fi
+usage
+
+# vim: et sw=2
```